cray_product_catalog/catalog_update.py

```diff
@@ -180,14 +189,17 @@ def update_config_map(data, name, namespace):
             # Key with same version exists in ConfigMap
             else:
                 if (merge_dict(data, product_data[PRODUCT_VERSION]) == product_data[PRODUCT_VERSION]
-                        and (current_version_is_active(product_data) or not SET_ACTIVE_VERSION)):
+                        and (current_version_is_active(product_data) or not SET_ACTIVE_VERSION)
```

Comment:
Eli, I applied some Boolean Algebra to understand this logic. I think it exposes a flaw.
Reference: https://grace.bluegrass.kctcs.edu/~kdunn0001/files/Simplification/4_Simplification_print.html
Note: 
\+ == OR
() == AND
!  == NOT

A  == current version is active
!B == (SET_ACTIVE_VERSION = None)
C  == There is no active version.
!D  == (REMOVE_ACTIVE_VERSION = None)

Equation as you have it in lines 192 and 193.
(A + !B)(C + !D) = AC + A!D + !BC + !B!D

AC can never be true, so it does not aid the equation.
A!D = This is possible, but just because !D doesn't mean we need to set the current version.
!BC = This is possible, but just because !B doesn't mean there are no active versions.
!B!D = This is possible and retains the status quo.

Reply:
Just to be clear this `if` statement is evaluating whether we `break` out of the loop because the desired state has already been reached.

Here's how I am interpreting the 3 cases that are OR'ed together (just disregarding AC for the same reason you gave).

A!D = Current version is active and we are not removing active versions
!BC = We don't want to set active version and there is no active version
!B!D = We don't care what is active

Also, just a quick clarification please:

> !BC = This is possible, but just because !B doesn't mean there are no active versions.

A little confused by that - C does mean there are no active versions.

This is a fairly problematic if statement though. I may need a moment to think through it all the way. Trying to figure out what combination of A-D would lead to undesirable behavior.

Reply:
@jsollom-hpe - do you think it would at least be more readable written like this?

```
                # Data to insert matches data found in configmap.
                if merge_dict(data, product_data[PRODUCT_VERSION]) == product_data[PRODUCT_VERSION]:
                    if SET_ACTIVE_VERSION and REMOVE_ACTIVE_FIELD:
                        # This should not happen (see main method).
                        raise SystemExit(1)
                    elif SET_ACTIVE_VERSION:
                        if current_version_is_active(product_data):
                            LOGGER.info("ConfigMap data updates exist; Exiting.")
                            break
                    elif REMOVE_ACTIVE_FIELD:
                        if not active_field_exists(product_data):
                            LOGGER.info("ConfigMap data updates exist; Exiting.")
                            break
                    else:
                        LOGGER.info("ConfigMap data updates exist; Exiting.")
                        break
```

Reply:
Moving forward with the rewrite as written above (d96f6fa). I think it is roughly equivalent to what we had before (though now there's a redundant check that SET_ACTIVE_VERSION and REMOVE_ACTIVE_FIELD are not both set). It is at least a bit more readable this way.

Reply:
Jason and I talked about this last week and agreed to move forward with this version.